### PR TITLE
Extract installer data using direct stream-stream copies

### PR DESCRIPTION
**Pull request by pchote**
The old implementation would read all the data into a memory buffer and then dump that buffer to disk as two separate options.  This is bad for the UI (no way to measure progress) and for memory (because we need to load entire files into memory – potentially twice).

This removes the intermediate memory buffers and adds (XX%) indicators to the installation progress UI.

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs [hunk 1]
```diff
@@ -226,9 +234,25 @@ void InstallFromDisc(string path, ModContent.ModDisc disc)
...
 		}
 
+		static void CopyStream(Stream input, Stream output, long length, Action<long> onProgress = null)
+		{
+			var buffer = new byte[4096];
+			var copied = 0L;
+			while (copied < length)
+			{
+				var n = (int)Math.Min(buffer.Length, length - copied);
+				input.Read(buffer, 0, n);
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs]: Need to check the actual bytes read here.

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs]: Good catch.  Fixed.

**Comment by pchote** [comment]
Added a new commit that adds a 25 MB size threshold for the % indicator.

**Comment by Phrohdoh** [comment]
Why 25MB?

**Comment by pchote** [comment]
It is a nice number that is larger than all the small files that we don't want the indicator on, but smaller than the large files that we do want it on. I wouldn't want to go much bigger for the sake of people stuck with spinning disks, or much smaller for the sake of people with SSDs.

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

**Comment by obrakmann** [comment]
Looks fine to me, :+1:

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5b236ab2ed4e93d51334fddba6d0771444cee0a4)

**Pull request merged**